annotated.go

```diff
@@ -272,210 +529,357 @@ func (la *lifecycleHookAnnotation) apply(ann *annotated) error {
+		if t == _typeOfContext && ctxPos < 0 {
+			ctxPos = i
+			continue
+		}
+		if isIn(t) {
+			fields := []reflect.StructField{_inAnnotationField}
+			for j := 1; j < t.NumField(); j++ {
+				field := t.Field(j)
+				if field.Type == _typeOfContext && ctxPos < 0 {
+					ctxStructPos = i
+					ctxPos = j
+					continue
+				}
+				fields = append(fields, field)
+			}
+			invokeParamTypes = append(invokeParamTypes, reflect.StructOf(fields))
+			continue
 		}
+		invokeParamTypes = append(invokeParamTypes, origHookFnT.In(i))
 	}
+	invokeFnT := reflect.FuncOf(invokeParamTypes, []reflect.Type{}, false)
+	invokeFn := reflect.MakeFunc(invokeFnT, func(args []reflect.Value) (results []reflect.Value) {
+		lc := args[0].Interface().(Lifecycle)
+		args = args[1:]
+		hookArgs := make([]reflect.Value, origHookFnT.NumIn())
 
-	return
-}
+		hookFn := func(ctx context.Context) (err error) {
+			// If the hook function has multiple parameters, and the first
+			// parameter is a context, inject the provided context.
+			if ctxStructPos < 0 {
+				offset := 0
+				for i := 0; i < len(hookArgs); i++ {
+					if i == ctxPos {
+						hookArgs[i] = reflect.ValueOf(ctx)
+						offset = 1
+						continue
+					}
+					if i-offset >= 0 && i-offset < len(args) {
+						hookArgs[i] = args[i-offset]
+					}
+				}
+			} else {
+				for i := 0; i < origHookFnT.NumIn(); i++ {
+					if i != ctxStructPos {
+						hookArgs[i] = args[i]
+						continue
+					}
+					t := origHookFnT.In(i)
+					v := reflect.New(t).Elem()
+					for j := 1; j < t.NumField(); j++ {
+						if j < ctxPos {
+							v.Field(j).Set(args[i].Field(j))
+						} else if j == ctxPos {
+							v.Field(j).Set(reflect.ValueOf(ctx))
+						} else {
+							v.Field(j).Set(args[i].Field(j - 1))
+						}
+					}
+					hookArgs[i] = v
+				}
+			}
+			hookResults := origHookFn.Call(hookArgs)
+			if len(hookResults) > 0 && hookResults[0].Type() == _typeOfError {
+				err, _ = hookResults[0].Interface().(error)
+			}
 
-func (la *lifecycleHookAnnotation) resolveLifecycleParamField(
-	param reflect.Value,
-	n int,
-) (
-	value reflect.Value,
-) {
-	if param.Kind() == reflect.Struct {
-		if n <= param.NumField() {
-			value = param.FieldByName(fmt.Sprintf("Field%d", n))
+			return
 		}
-	}
+		lc.Append(la.buildHook(hookFn))
+		return
+	})
 
-	return value
+	installerType := reflect.FuncOf(append(resultTypes, paramTypes...), []reflect.Type{_typeOfError}, false)
+	hookInstaller = reflect.MakeFunc(installerType, func(args []reflect.Value) (results []reflect.Value) {
+		results = []reflect.Value{_nilError}
+		var scope *dig.Scope
+		switch la.Type {
+		case _onStartHookType:
+			scope = ann.container.Scope("onStartHookScope")
+		case _onStopHookType:
+			scope = ann.container.Scope("onStopHookScope")
+		}
+		ctor := makeHookScopeCtor(paramTypes, resultTypes, args)
+		if err := scope.Provide(ctor); err != nil {
+			results[0] = reflect.ValueOf(fmt.Errorf("Error providing possible parameters for hook installer: %w", err))
```

Comment:
Error strings should be lowercase. 
https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Reply:
Fixed